### PR TITLE
Require that public TLS and SSH keys are provided to register via token

### PR DESCRIPTION
**Pull request by zmb3**
The original behavior attempted to make providing public keys optional,
and would generate keys if they were not provided. This had several
problems:

- The auth server is generating private keys for nodes and is
  potentially able to share them over the network.
- The return value for keys.Key would sometimes be set and sometimes
  be empty (the key is only set if the auth server generated it and
  knows what the key is)
- We only ever relied on this behavior as a shortcut in test code.
  In the production code this behavior was never used (and actually
  never worked due to a bug that would overwrite and discard the
  generated private key)

This commit requires that public keys are always provided, ensuring
that the private key is generated locally and never known by the
auth server.

It also results in a cleaner error message when either or both of the
public keys are missing from the request.

### CODE REVIEW

File: lib/auth/auth.go [hunk 1]
```diff
@@ -1462,6 +1461,9 @@ func (req *GenerateServerKeysRequest) CheckAndSetDefaults() error {
 	if len(req.Roles) != 1 {
 		return trace.BadParameter("expected only one system role, got %v", len(req.Roles))
 	}
+	if len(req.PublicSSHKey) == 0 || len(req.PublicTLSKey) == 0 {
```
Comment (zmb3) [inline lib/auth/auth.go:1464]: Note to reviewers: the code always failed if this case were true, but it failed due to trying to parse an empty PEM block. Moving this check up into `CheckAndSetDefaults` just makes the error easier to diagnose.

File: lib/auth/auth.go [hunk 2]
```diff
@@ -1506,31 +1508,13 @@ func (a *Server) GenerateServerKeys(req GenerateServerKeysRequest) (*PackedKeys,
...
-			return nil, trace.Wrap(err)
-		}
-
-		// reuse the same RSA keys for SSH and TLS keys
-		cryptoPubKey, err = sshutils.CryptoPublicKey(pubSSHKey)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-
+	_, _, _, _, err = ssh.ParseAuthorizedKey(req.PublicSSHKey)
```
Comment (codingllama) [inline lib/auth/auth.go:1511]: nit: inline err in the if? Smaller scope and so on.

File: lib/auth/auth.go [hunk 3]
```diff
@@ -1560,7 +1544,8 @@ func (a *Server) GenerateServerKeys(req GenerateServerKeysRequest) (*PackedKeys,
 			return nil, trace.BadParameter("failed to load host CA for %q: %v", clusterName.GetClusterName(), err)
 		}
 		if !req.Rotation.Matches(ca.GetRotation()) {
-			return nil, trace.BadParameter("the client expected state is out of sync, server rotation state: %v, client rotation state: %v, re-register the client from scratch to fix the issue.", ca.GetRotation(), req.Rotation)
+			return nil, trace.BadParameter("the client expected state is out of sync, server rotation state: %v, "+
+				"client rotation state: %v, re-register the client from scratch to fix the issue.", ca.GetRotation(), req.Rotation)
```
Comment (codingllama) [inline lib/auth/auth.go:1548]: Suggestion: I fancy doing it like this, so the text starts at the same column. Yours is fine too.

```suggestion
			return nil, trace.BadParameter("" + 
				"the client expected state is out of sync, server rotation state: %v, "+
				"client rotation state: %v, re-register the client from scratch to fix the issue.",
				ca.GetRotation(), req.Rotation)
```

File: lib/auth/helpers.go [hunk 4]
```diff
@@ -377,9 +378,33 @@ func (a *TestAuthServer) GenerateUserCert(key []byte, username string, ttl time.
...
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	tlsPublicKey, err = tlsca.MarshalPublicKeyFromPrivateKeyPEM(sshPrivate)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	return tlsPublicKey, publicKey, nil
```
Comment (codingllama) [inline lib/auth/helpers.go:392]: It looks like publicKey enters and leaves unchanged? Drop both the parameter and the return?

File: lib/auth/helpers.go [hunk 5]
```diff
@@ -840,14 +867,27 @@ func (t *TestTLSServer) Stop() error {
 
 // NewServerIdentity generates new server identity, used in tests
 func NewServerIdentity(clt *Server, hostID string, role types.SystemRole) (*Identity, error) {
+	private, public, err := clt.GenerateKeyPair("")
```
Comment (codingllama) [inline lib/auth/helpers.go:870]: Suggestion: for consistency?

```suggestion
	priv, pub, err := clt.GenerateKeyPair("")
```

**Review by codingllama — APPROVED** [review]
LGTM.

File: integration/integration_test.go [hunk 6]
```diff
@@ -222,7 +222,7 @@ func testAuditOn(t *testing.T, suite *integrationTestSuite) {
 	tr := utils.NewTracer(utils.ThisFunction()).Start()
 	defer tr.Stop()
 
-	var tests = []struct {
+	tests := []struct {
```
Comment (zmb3) [inline integration/integration_test.go:225]: These changes were applied automatically via gofumpt - happy to revert if anyone prefers the longer var declaration.

**Review by rosstimothy — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Based off our discussion, I don't think this is an issue, but if you plan to backport this to `branch/v7` please test and make sure we didn't break 6.2 clients (our own + api clients).

File: lib/auth/auth.go [hunk 7]
```diff
@@ -1506,31 +1508,12 @@ func (a *Server) GenerateServerKeys(req GenerateServerKeysRequest) (*PackedKeys,
...
-		}
-
-		// reuse the same RSA keys for SSH and TLS keys
-		cryptoPubKey, err = sshutils.CryptoPublicKey(pubSSHKey)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-
+	if _, _, _, _, err := ssh.ParseAuthorizedKey(req.PublicSSHKey); err != nil {
+		return nil, trace.BadParameter("failed to parse SSH public key")
```
[the remark this answers is not in the document]
Comment (zmb3) [reply lib/auth/auth.go:1512]: Only because the original code wasn’t.

**Pull request merged**